fast_llm/engine/training/config.py

```
@@ -349,30 +338,23 @@ class TrainerConfig(PretrainedFastLLMModelConfig, ExperimentConfig):
     _abstract = True
     # TODO: Generalize data, schedule, logging, etc.
     training: TrainingConfig = Field(
-        default_factory=TrainingConfig,
```

you mentioned in the other PR that this is necessary now, but I don't quite understand why. if you've got several classes of type `TrainingConfig`, you want to have a straightforward way to say which one is the default. that's pretty much it

---

Problem was all sorts of potential issues happening if not going through `_from_dict`, not necessarily related to this PR (ex. dataclasses doing the wrong thing when calling it for nested configs). But now that I enforced going through it (with the metaclass `__call__`), I think I can safely bring it back.

---

Actually this won't work without some hacking around. Problem is we don't just want to set a default, we want to set a default to the nested config's `type` field even is a value is provided. And we can't just set a default to `type` because it would propagate to subclasses and cause all sorts of trouble.

---

Ok, though we still need a clean, consistent way to specify default values for config fields.

```python
field: SomeConfig = Field(default_factory=SomeConcreteConfig)
```
should just work. If `SomeConcreteConfig` is a subclass in dynamic dispatch / tagged union style setup, then instantiating it directly should still produce a valid config object. The fact that it has a type or tag field shouldn't block the default from being usable. If it's not currently reliable, that's a sign that we're over-complicating the dispatch mechanism than a reason to give up on defaults.

---

This would work perfectly fine, just not in a useful way, and it would be a ton of trouble.

`default_factory` (and `default`) is for when a value is *not provided*. This means it is (and should be) ignored whenever anything is passed, including an empty dict. So it's not suitable for setting the default field type, because setting any value will undo it. 

On top of that, the config mechanism doesn't distinguish whether nested fields are passed explicitly or just empty. So some defaults may lead to incorrect serialization paths, ex: `{} --(instantiate)-> Config({}) --(serialize)->  [MISSING] --(instantiate)->  Config([DEFAULT])`. (in fact this happens for all nested fields, not just configs, but configs are the biggest risk)

---

Also note that this doesn't really have to do with dynamic classes or the present PR. It's an instantiation thing, and setting non-dynamic defaults would lead to the same problem, ex: `[Field(default_factory=lambda: SomeConcreteConfig](field: SomeConfig = Field(default_factory=lambda : SomeConcreteConfig(some_field=3))` would have the same problem. This wasn't working before either, all I did here is to replace the silent errors with explicit checks.

---

Hm, the issue you're describing is surfacing because of how dispatch and serialization are currently wired. Let's look at the example you gave:

```
{} --(instantiate)-> Config({}) --(serialize)-> [MISSING] --(instantiate)-> Config([DEFAULT])
```

This only becomes a problem because the **serializer aggressively drops fields** when their values match defaults. In this case, it drops the *entire field*, including the tag, just because the config looks "default-ish." That's the actual cause of round-trip inconsistencies, and not the use of `default_factory`!

We should not be optimizing serialized configs for brevity at the cost of correctness or developer sanity. I've raised this before, many times actually, and I'll say it again: **nobody cares about shaving lines off a YAML file**, especially if it leads to complexity and development headaches.

What we need instead is what every normal config system does:

* If the field is missing or explicitly set to `{}`, call `default_factory()` and use the result.
* If a dict is provided without a tag, use `default_factory()` and apply the provided keys as overrides.
* If a dict includes a tag, use that to resolve the correct subclass and instantiate it with the overrides as usual.

This works the same for dynamic and non-dynamic configs. It produces clean, predictable defaults, supports partial overrides, and round-trips cleanly — **as long as we stop deleting structure during serialization**.

If our config system can't support this behavior without hacks, that's a design flaw in the dispatch/serialization layer and not a reason to abandon reasonable defaulting behaviour.

---

Let's not get sidetracked with what we serialize. As I mentioned multiple times already, **I do care a lot about it**, clear and concise configs are essential to me for development, testing, debugging, understanding bug reports, etc. There are plenty of options to get the whole config when you want it. And as for the issue mentioned here, they are easily fixable, but it's not urgent.


> What we need instead is what every normal config system does:
> 
> * If the field is missing or explicitly set to `{}`, call `default_factory()` and use the result.
> * If a dict is provided without a tag, use `default_factory()` and apply the provided keys as overrides.
> * If a dict includes a tag, use that to resolve the correct subclass and instantiate it with the overrides as usual.

Afaik that's **not an established standard**.There are two perfectly good ways to see defaults for config:
1. Config as arguments: Defaults are used when no value is provided. This is the standard for dataclasses, pydantic `BaseModel`, and coding in general.
2. Config as update: The default determines a baseline, which the config updates in a nested way. This is the standard for omegaconf/hydra.

As of now, Fast-LLM lies deeply in option 1. Changing this is not impossible but would be complex and have deep implications, and the discussion on this lies far outside the scope of this PR. 

Either way, switching to option 2 would not help with the present problem of selecting a default subclass, in fact it would make things worse. Because once we pick and create a subclass (say B) for a field (of type A), we can't swap with any subclass C of A, because we committed to defining the fields of class B, i.e. we C must be a subclass of B. (This is what omegaconf does.) So we're not setting a default subclass but rather overriding the type hint, so we might as well just set the type hint right to begin with. In option 1 we don't have this problem because we use the actual type from the beginning.

What we actually want is a default for the `type` field, ant it's just like any field default but needs a slightly different handling since it needs to be set prior to instantiation.

---

I've spent many hours this week diving into this PR and its discussions to understand the broader impact. I want to be crystal-clear about where we stand:

**Fast-LLM is an model training framework, not a configuration R&D project.**

Yet the config subsystem keeps burning days of senior time. That's time we're not spending on training models, fixing bugs, or shipping features. We've now burned more than a week on a problem Pydantic solves out of the box. That's not sustainable.

Every extra layer of custom logic raises the bar for contributors, concentrates knowledge in one person, and stalls the real work: training models and shipping features. As the team lead, I have to call it what it is: it's an *operational risk.*

We need to reverse this pattern. The config system should get out of our way, not pull the whole team into its gravity well.

And no, this isn't about shutting down ideas. It's about protecting this team from complexity creep that slows everyone down.

> As I mentioned multiple times already, I do care a lot about it, clear and concise configs are essential to me for development, testing, debugging, understanding bug reports, etc.

This illustrates the underlying issue: the current design optimizes for what the framework maintainer finds convenient for himself, but not for what contributors actually need when developing models or features.

Now, let's look at what works and doesn't work in Pydantic:

```python
from pydantic import BaseModel, Field, ValidationError
from typing import Literal, Union

class Foo(BaseModel):
    type: Literal['foo'] = 'foo'
    x: int = 1

class Bar(BaseModel):
    type: Literal['bar'] = 'bar'
    y: int = 2

Item = Union[Foo, Bar]

class Wrapper(BaseModel):
    item: Item = Field(default_factory=Foo, discriminator='type')

print(Wrapper().model_dump())  # works, default_factory runs, uses Foo()
# print(Wrapper(item={}))  # doesn't work, validation fails because of missing discriminator
# print(Wrapper(item={'x': 42}))  # doesn't work, validation fails because of missing discriminator
print(Wrapper(item={'type': 'bar', 'y': 99}).model_dump())  # works, dispatches to Bar
```

This setup gets most things right:

* Dynamic dispatch is opt-in.
* The discriminator is explicit and user-controllable.
* Defaults work when the field is absent.
* There is no magic registry.

Does it handle every edge case? No. But it works well for the common case. And that's the bar Fast-LLM needs to meet.

Right now, we're debating edge cases that haven't materialized (e.g., sub-subclassing edge-cases should be rare, and if we're running into them, that's a sign to simplify the class hierarchy and not to complicate things further.) Meanwhile, the team is stuck. That's the wrong tradeoff.

In hindsight, we should have adopted Pydantic from the start. But since we're here now, and deadlines are real, we need to course-correct pragmatically.

We've seen the cost of keeping this overly flexible. It's time to simplify. Here are the principles we need to lock in now:

1. Dynamic config behaviour must be opt-in, via a mixin or decorator. Most config classes will just be plain config dataclasses.

2. Subclasses declare their own tag via something, a `ClassVar[str]`, or decorator param. It doesn't really matter. Your choice, @jlamypoirier. But it must be self-contained and discoverable.

3. If a config field is missing, we use `default_factory`. If a dict is provided without a tag (it shouldn't), we either:
    * inject the default tag automatically (preferred), or
    * error clearly with a message explaining the missing tag (acceptable fallback).

4. No global subclass registries tied to CLI routing. CLI dispatch should be its own, explicit layer. Config structure should not bend to command-line ergonomics.

These constraints exist to make Fast-LLM maintainable and usable by the whole team. The bar is not "can @jlamypoirier make this work." It's "can any contributor pick this up, understand it, and extend it easily."

That's what makes this framework sustainable.

---

Let's be clear, the time we're taking is due to the overly detailed discussion we're having here. I had a working solution 2 weeks ago, with near-zero complexity on the user side. 

> We've seen the cost of keeping this overly flexible. It's time to simplify. Here are the principles we need to lock in now:
> 
> 1. Dynamic config behaviour must be opt-in, via a mixin or decorator. Most config classes will just be plain config dataclasses.

It's already opt-in, there is nothing to do when we don't want it, and nothing will happen if we don't use it. Everything else is implementation details. 

> 2. Subclasses declare their own tag via something, a `ClassVar[str]`, or decorator param. It doesn't really matter. Your choice, @jlamypoirier. But it must be self-contained and discoverable.

Already done, see the decorator implementation.

> 3. If a config field is missing, we use `default_factory`. If a dict is provided without a tag (it shouldn't), we either:
>    
>    * inject the default tag automatically (preferred), or
>    * error clearly with a message explaining the missing tag (acceptable fallback).

That's already what we're doing.

> 4. No global subclass registries tied to CLI routing. CLI dispatch should be its own, explicit layer. Config structure should not bend to command-line ergonomics.

It is and always had been the opposite, the cli adapts to the config system. I updated the cli system according to what the config sytstem now supports, and your request to preserve the syntax

---

Thanks, @jlamypoirier. I want to respond clearly to your points:

> "I had a working solution 2 weeks ago..."

The PR is still in draft, tests were failing non-stop until the last commit, and key review feedback -- including on default handling, tagging, and registration -- remains unresolved or blatantly ignored. There was no point at which this was merge-ready. It stalled until I pushed.

> "There is nothing to do when we don't want [dynamic dispatch]..."

That's not what "opt-in" means. Right now, every config class implicitly enters a world with base-class registries, tag handling, and dynamic behaviours -- even if it doesn't use them. Every single config class pays this complexity tax. That's *passive inclusion*, not opt-in.

What I've asked for is explicit control: if a developer doesn't add a mixin or decorator, the config class behaves like a plain (config) dataclass. No registry, no tagging, no indirection. That's what "opt-in" means in a maintainable system. And I had described how this can work eons ago: https://github.com/ServiceNow/Fast-LLM/pull/104#issuecomment-2599496346

> "Already done..." / "That's already what we're doing."

If that were the case, this conversation wouldn't be happening. Are the default factories back? No? Well then.

> “The CLI adapts to the config system..."

We shouldn't even be debating this. Why is there even coupling between the CLI commands and the config system? We have 3 or 4 commands right now. Give them one, hard-coded root config class each and call it a day.

We're deep in this because we didn't push for simplicity early. I'm asking for a course correction now, not to block you, but to unblock everyone else.

---

> That's not what "opt-in" means. Right now, every config class implicitly enters a world with base-class registries, tag handling, and dynamic behaviours -- even if it doesn't use them. Every single config class pays this complexity tax. That's _passive inclusion_, not opt-in.
> 
> What I've asked for is explicit control: if a developer doesn't add a mixin or decorator, the config class behaves like a plain (config) dataclass. No registry, no tagging, no indirection. That's what "opt-in" means in a maintainable system. And I had described how this can work eons ago: [#104 (comment)](https://github.com/ServiceNow/Fast-LLM/pull/104#issuecomment-2599496346)

Let's not debate about semantics. Nothing visible happens if the developer doesn't do anything, what happens in the background is irrelevant. Unless I'm missing something?

> If that were the case, this conversation wouldn't be happening. Are the default factories back? No? Well then.

Default factories were never there for config classes. All I can do at this time is force them to be an exact match with the type hint, which is equivalent to them not being there. Anything more is not an option.

> We shouldn't even be debating this. Why is there even coupling between the CLI commands and the config system? We have 3 or 4 commands right now. Give them one, hard-coded root config class each and call it a day.

I don't get what's wrong here. I can revert the changes if you want. But we have a perfect match between cli commands and the new dynamic config system, it would be weird to keep and maintain a separate infrastructure just because we want to.

---

Thanks for the update, @jlamypoirier. I don't want to drag this out further, so we'll move forward for now.

That said, I want to clearly record the unresolved concerns:

Every config class now implicitly participates in dynamic dispatch and registry logic, even when that complexity isn't needed. The cognitive overhead of the config system remains too high, and this PR made it worse. The sheer review burden here shows how far the design diverges from what most developers expect in a config system. We've burned over a week of senior time debating mechanics, and that time we could have been spent on models, training, or features.

We continue to build and maintain a bespoke config layer, which distracts from model development. A better choice would be to lean on mature solutions like Pydantic, which are battle-tested and solve 95% of our use cases with far less friction.

If the complexity tax of the current system continues to stall progress, I will propose a full migration to Hydra + OmegaConf + Pydantic. 

For now, I'm prioritizing unblocking the team and moving forward.